src/topology/builder.rs

```diff
@@ -51,23 +51,27 @@ pub async fn build_pieces(
     // Build sources
     for (name, source) in config
         .sources
-        .iter()
+        .iter_mut()
         .filter(|(name, _)| diff.sources.contains_new(&name))
     {
         let (tx, rx) = futures::channel::mpsc::channel(1000);
         let pipeline = Pipeline::from_sender(tx, vec![]);
 
-        let typetag = source.source_type();
+        let typetag = source.config.source_type();
 
         let (shutdown_signal, force_shutdown_tripwire) = shutdown_coordinator.register_source(name);
 
-        let context = SourceContext {
-            name: name.into(),
-            globals: config.global.clone(),
-            shutdown: shutdown_signal,
-            out: pipeline,
-        };
-        let server = match source.build(context).await {
+        let context = SourceContext::new(
+            name,
+            source.identifier.clone(),
+            config.global.clone(),
+            shutdown_signal,
+            pipeline,
+        );
+        if source.identifier.is_none() {
+            source.identifier = Some(context.identifier.clone());
+        }
```

Comment:
Just checking, the expectation here is that the id will be `None` for components that did not match the name of an existing component and get an id copied over in `reload_config_and_respawn`?

Again, it'd be nice to have a little clearer invariants around when this field is expected/required to be present but I don't have a great suggestion on how to do that.